### ISSUE TITLE
Type narrowing with type()

### ISSUE THREAD

**Issue by adamhl8** [microsoft/pyright#4719]
I have a function structured like this:
```python
def my_func(val: str | list[str]):
    # If val is a str, place into a list
    if type(val) == str:
        val = [val]
```

However, inside the `if`, the type of `val` isn't being narrowed to `str`, so pyright (correctly) reports the following:
```
Expression of type "list[str | list[str]]" cannot be assigned to declared type "str | list[str]"
  Type "str | list[str]" cannot be assigned to type "str"
    "list[str]" is incompatible with "str"
```

I can instead do:
```python
def my_func(val: str | list[str]):
    if isinstance(val, str):
        val = [val]
```
This works as expected.

Why does `isinstance()` narrow the type of `val` but not `type()`?

**Comment by erictraut** [comment]
The expression `type(x) == y` is not currently a supported type guard pattern. For a complete list of supported patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards). Each pattern requires its own logic and potentially slows down type analysis, so we add support for only those patterns that are commonly used.

The pattern you're using here is somewhat uncommon. I've rarely seen it used. The more idiomatic pattern is `type(x) is y`, and pyright does support this pattern currently.

That said, it appears that mypy supports the patterns `type(x) == y` and `type(x) != y` for type guards, so it makes sense for pyright to support this as well. I've added this support, and it will appear in the next release.

**Comment by adamhl8** [comment]
Awesome, thank you for the quick response.

**Comment by adamhl8** [comment]
@erictraut This is sort of unrelated, but I'm a bit confused about the behavior of type guards and the effect they have on types outside their scope. Reading the section on [Type Narrowing](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing), it sounds like types will only be narrowed within the `if` block.

When using `type(val) is str`, the type of `val` is narrowed inside the `if` block but not outside. 
```python
def my_func(val: str | list[str]):
    # If val is a str, place into a list
    if type(val) is str:
        val = [val]
        reveal_type(val)  # list[str]
        
    reveal_type(val)  # list[str] | str
```

When using `isinstance(val, str)`, the type of `val` is remains narrowed.
```python
def my_func(val: str | list[str]):
    # If val is a str, place into a list
    if isinstance(val, str):
        val = [val]
        reveal_type(val)  # list[str]
        
    reveal_type(val)  # list[str]
```

FWIW, it seems strange to me that in the first case the type of `val` doesn't remain narrowed to `list[str]`. It is impossible for `val` to not end up as `list[str]` in that case.

**Comment by erictraut** [comment]
Type narrowing isn't limited to the block guarded by a conditional. It persists beyond the conditional block. Consider the following:

```python
def func(x: str | int):
    reveal_type(x) # str | int
    if isinstance(x, str):
        print('str')
    else:
        return
    reveal_type(x) # str
```

Some type guards allow for both "positive" (if) and "negative" (else) narrowing. The `isinstance` guard is an example of this, but the `type(x) is y` guard supports only positive narrowing. Negative narrowing isn't possible for `type(x) is y` because the annotation `str` means "an instance of `str` or a subtype thereof", so `type(val) is str` could return false if `val` was a subtype of `str`.

Your code may not contain an explicit `else` statement, but there's an implied `else: pass`. The effective type after the `if` / `else` statement is the union of the two narrowed types (positive and negative).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.297, which I just published. It will also be included in a future release of pylance.
